### ISSUE TITLE
Superpose screens out &stdlib type definitions in imported files

### ISSUE THREAD

**Issue by patham9** [trueagi-io/hyperon-experimental#396]
The following code snippet:
```
(= (crash $C) (superpose ((if (> $C 0) (crash (- $C 1)) nocrash))))
!(crash 1)
```
gives "nocrash" as expected, but when the first statement is in test.metta, then the following code causes a segfault:
```
!(import! &self test.metta)
!(crash 1)
```

**Comment by Necr0x0Der** [comment]
Some notes:
* Tokenizer works correctly with imports, and arithmetics is performed correctly, so `(- $C 1)` is reduced and calculated.
* `(> $C 0)` is compared correctly.
* If we remove `superpose`, `(= (crash $C) (if (> $C 0) (crash (- $C 1)) nocrash))` works correctly even when imported.

How does it happen that the recursion is not stopped? The problem is that all arguments of `if` are being evaluated in this specific case. Normally, they are not evaluated because `if` is typed as `(: if (-> Bool Atom Atom $t))` - see https://github.com/trueagi-io/hyperon-experimental/blob/27861e63af1417df4780d9314eaf2e8a3b5cde06/lib/src/metta/runner/stdlib.rs#L1187
This type definition is added to the main mette script by inserting the reference to `&stdlib` space into it.
Thus, `superpose` screens out the reference to `&stdlib` code from the imported file (it is also imported not by adding its content to `&self`, but by inserting a reference to it).

Basically, `SuperposeOp` is created
https://github.com/trueagi-io/hyperon-experimental/blob/27861e63af1417df4780d9314eaf2e8a3b5cde06/lib/src/metta/runner/stdlib.rs#L1135
with the space corresponding only to the imported file. `ImportOp` doesn't insert `&stdlib` into the imported space to avoid double definitions (and double evaluations) - https://github.com/trueagi-io/hyperon-experimental/blob/27861e63af1417df4780d9314eaf2e8a3b5cde06/lib/src/metta/runner/stdlib.rs#L59 . Then, `SuperposeOp` interprets the given expression in its own context space: https://github.com/trueagi-io/hyperon-experimental/blob/27861e63af1417df4780d9314eaf2e8a3b5cde06/lib/src/metta/runner/stdlib.rs#L655 , and when this is the space of imported script, it doesn't contain `&stdlib`.

**Comment by Necr0x0Der** [comment]
There are two possible solutions on the `superpose` side:
* somehow pass main space to `superpose`
* temporarily insert `&stdlib` into space, for which `superpose` is executed

A hotfix solution for this particular example is just to add `(: if (-> Bool Atom Atom $t))` in `test.metta`, if it is really needed.

I'm not sure if we want to fix this on the `superpose` side in the current implementation, or we want to wait for Minimal MeTTa or/and a redesigning of imports and libraries in general.

**Comment by patham9** [comment]
Thank you having looked into this!
I agree for this simplified example, but the additional statement does not fix the issue for my more complex case.
But no worries it is not urgent for me to have this fixed, as this will suffice for me:
`cat file1.metta file2.metta > RUN.metta && metta RUN.metta`

**Comment by Necr0x0Der** [comment]
An example of a more complex case could be helpful to avoid such issues

**Comment by patham9** [comment]
Here is another:

```
!(import! &self **nal.metta**)
!(|- ((Garfield --> cat) (1.0 0.9)) ((cat --> animal) (1.0 0.9)))
thread '<unnamed>' panicked at 'already mutably borrowed: BorrowError', /Users/paha9538/hyperon-experimental/lib/src/space/mod.rs:296:34
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: Rust panics must be rethrown
Abort trap: 6

```
**nal.metta:**
https://gist.github.com/patham9/582e45c3fab4af85df7f883a972c46e3

**Comment by patham9** [comment]
I had another one, but I posted the wrong part, sorry one sec.
